opencensus/metrics/registry.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2019, OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime
+import threading
+
+from opencensus.metrics.export import metric_producer
+
+
+class Registry(metric_producer.MetricProducer):
+    """A collection of gauges to be exported together.
+
+    Each registered gauge must have a unique `descriptor.name`.
+    """
+
+    def __init__(self):
+        self.gauges = {}
+        self._gauges_lock = threading.Lock()
+
+    def __repr__(self):
+        return ('{}(gauges={}'
+                .format(
+                    type(self).__name__,
+                    self.gauges
+                ))
+
+    def add_gauge(self, gauge):
+        """Add `gauge` to the registry and return it.
+
+        Raises a `ValueError` if another gauge with the same name already
+        exists in the registry.
+
+        :type gauge: class:`opencensus.metrics.export.gauge.LongGauge` or
+        class:`opencensus.metrics.export.gauge.DoubleGauge`
+        :param gauge: The gauge to add to the registry.
+
+        :rtype: class:`opencensus.metrics.export.gauge.LongGauge` or
+        class:`opencensus.metrics.export.gauge.DoubleGauge`
+        :return: The gauge that was added to the registry.
+        """
+        if gauge is None:
+            raise ValueError
+        name = gauge.descriptor.name
+        with self._gauges_lock:
+            if name in self.gauges:
+                raise ValueError(
+                    'Another gauge named "{}" is already registered'
+                    .format(name))
+            self.gauges[name] = gauge
+
+    def get_metrics(self):
+        """Get a metric for each gauge in the registry at the current time.
+
+        :rtype: set(:class:`opencensus.metrics.export.metric.Metric`)
+        :return: A set of `Metric`s, one for each registered gauge.
+        """
+        now = datetime.now()
+        metrics = set()
+        with self._gauges_lock:
```

Comment:
Depending on whether it is CPython or something else, we might not need the threading.Lock while getting `values()`.

Reply:
What about if another thread changes a value while we're iterating here? E.g.

```python
import time
from concurrent.futures import ThreadPoolExecutor

tpe1 = ThreadPoolExecutor(max_workers=1)
tpe2 = ThreadPoolExecutor(max_workers=1)

dd = dict(zip(range(10), range(10)))

def read_values():
    global dd
    for vv in dd.values():
        print(vv, end=", ")
        time.sleep(.01)
    print()
        
def mess_with_values():
    global dd
    for kk in dd:
        dd[kk] = -dd[kk]
        time.sleep(.01)


tpe1.submit(read_values)
tpe2.submit(mess_with_values)
```
```
0, 1, 2, 3, 4, -5, -6, -7, -8, -9, 
```

Reply:
I think there are better ways to do this, all the locking in gauges so far has been pretty naive. One obvious improvement is to use a read/write lock to let more than one thread read at once. I didn't spend time optimizing this yet because I don't think we're likely to have a bottleneck here, but I did want to make sure the class was threadsafe.

Reply:
I thought we are only adding new gauges (add_gauge) instead of altering the existing gauges?
And after we take a snapshot of the registered gauges, we get the metrics based on a pre-calculated value `now`.

Reply:
I thought we'd want to remove gauges too, but after talking to @bogdandrutu it turns out that this is by design.

Looking more closely at `dictview` I think you're right that we don't need a lock here.